admin can handle bottom nav visibility

Nikunjsharma0 opened the pull request:
to control the visibility of bottom nav items

Nikunjsharma0 committed:
added bottonnavVisiblity flag in siteEntity
to control the visibility of bottom nav items

Nikunjsharma0 committed:
BottomNavVisibilityFlag added

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
Merge branch 'dev-simplified-onboarding' into dev-simplified-onboarding-edit-bottomnav-adminsetting

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
Merge branch 'primary' into dev-simplified-onboarding-edit-bottomnav-adminsetting

Nikunjsharma0 committed:
Merge branch 'primary' into dev-simplified-onboarding-edit-bottomnav-adminsetting

Nikunjsharma0 committed:
added BOTTOM_NAV_DEFAULT_FLAG as 15L

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
Merge branch 'dev-simplified-onboarding-edit-bottomnav-adminsetting' into primary-bottomnav-visiblitysetting

# Conflicts:
#	lib-database/schemas/com.ustadmobile.core.db.UmAppDatabase/209.json
#	lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/ext/UmAppDatabaseMigrations.kt
#	lib-database/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/Site.kt

Nikunjsharma0 committed:
commit

mikedawson commented on lib-database/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/Site.kt:
```diff
@@ -60,10 +64,15 @@ class Site {
 
     var authSalt: String? = null
 
+    var bottomNavVisibilityFlag: Long? = null
```
Add ColumnInfo annotation (including default value) - you should use the real default value.

Nikunjsharma0 committed:
changing MIGRATION

Nikunjsharma0 committed:
added for web too

poojaustad committed:
End to end tests added for admin manage bottom navigation bar- feature

poojaustad committed:
Merge remote-tracking branch 'origin/primary-bottomnav-visiblitysetting' into primary-bottomnav-visiblitysetting

poojaustad committed:
edits on toggle switches

Nikunjsharma0 committed:
test tag added

Nikunjsharma0 committed:
Merge remote-tracking branch 'origin/primary-bottomnav-visiblitysetting' into primary-bottomnav-visiblitysetting

Nikunjsharma0 committed:
migration added

Nikunjsharma0 committed:
test id added

poojaustad committed:
added id's to toggles

poojaustad committed:
changed visible to exist

poojaustad committed:
changed visible to exist

Nikunjsharma0 committed:
test id and migration changes

Nikunjsharma0 committed:
Merge remote-tracking branch 'origin/primary-bottomnav-visiblitysetting' into primary-bottomnav-visiblitysetting

poojaustad committed:
Added id's for switches

mikedawson commented on lib-database/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/Site.kt:
```diff
@@ -60,10 +65,16 @@ class Site {
     var authSalt: String? = null
 
+    @ColumnInfo(defaultValue = "0")
+    var bottomNavVisibilityFlag: Long = 0
```
Constants for flags should be on this companion object

mikedawson replied:
Should be an int - 32 potential values is more than enough

mikedawson replied:
Kept as long because other functions to handle permission constants etc use Long type.

mikedawson commented on lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/BottomNavVisibilityFlags.kt:
```diff
@@ -0,0 +1,10 @@
+package com.ustadmobile.core.db
+
+object BottomNavVisibilityFlags {
```
This class should be removed - flags should be in Site.kt

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/app/App.kt:
```diff
@@ -54,12 +58,30 @@ import moe.tlaster.precompose.navigation.rememberNavigator
+    bottomNavVisibilityFlag: Long?
+): List<TopNavigationItem> {
+    return APP_TOP_LEVEL_NAV_ITEMS.filter {
+        when (it.destRoute) {
```
Flag should be added to the data class TopNavigationItem

mikedawson replied:
Please apply the same approach on the web version

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/Sidebar.kt:
```diff
@@ -32,14 +34,30 @@ import web.cssom.px
     val icon: FC<*>? = null,
 )
-
+fun getVisibleRootScreens(bottomNavVisibilityFlag: Long?): List<RootScreen> {
```
Should use same approach as per comment on App.kt (compose)

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/UstadScreens.kt:
```diff
@@ -100,6 +102,13 @@ val UstadScreens = FC<Props> {
+    val currentDb: UmAppDatabase =  loaderData.di.direct.on(accountManager.activeLearningSpace)
+        .instance(tag = DoorTag.TAG_DB)
+
+    val currentSite by currentDb.siteDao().getSiteAsFlow().collectAsState(null)
```
Pass the current site as a parameter (like app.kt) not the screen list

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/site/edit/BottomNavPermissionConstants.kt:
```diff
@@ -0,0 +1,14 @@
+
+object BottomNavPermissionConstants {
+
+    val BOTTOM_NAV_LABELS = listOf(
```
Should be removed as per comments on flags

mikedawson replied:
Add a new file: SiteConstants (in com.ustadmobile.viewmodel.site ). 

That should have a list of the top level items info that is common (e.g. the route name, the label string constant, the flag constant).

Then on each platform - create a list of those common constants plus the icon 

```
val iconMap = mapOf(
 flag to icon
) 
val TOP_LEVEL = commonTopLevel.map { ...
  it to iconMap[it.flag]
}
```

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/view/site/edit/SiteEditScreen.kt:
```diff
@@ -25,18 +29,20 @@ import react.FC
 
     val strings = useStringProvider()
-
+    val theme by useRequiredContext(ThemeContext)
```
The IDE is saying its unused - so it shouldn't be here. Don't ignore warnings on commit

mikedawson replied:
Still - has unused imports (can be removed by simply using ctrl+alt+O)

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/site/edit/SiteEditViewModel.kt:
```diff
@@ -158,7 +167,20 @@ class SiteEditViewModel(
-
+    fun onTogglePermission(flag: Long) {
+        _uiState.update { prev ->
+            val entityVal = prev.site
```
No need for this if/else - can and should be prev.site?.copy()

mikedawson commented on lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/ext/UmAppDatabaseMigrations.kt:
```diff
@@ -1692,15 +1693,23 @@ val MIGRATION_206_207 = DoorMigrationStatementList(206, 207) { db ->
 
+
 val MIGRATION_207_208 = DoorMigrationStatementList(207, 208) { db ->
+
```
This migration number needs to be bumped, and previous 207-208 migration should be here.

Nikunjsharma0 committed:
resolving pullrequest comments

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
cypress.config.js added

poojaustad committed:
cypress config added

poojaustad committed:
cypress added test if only one option in navigation bar is enabled

poojaustad committed:
maestro added test if only one option in navigation bar is enabled

Nikunjsharma0 committed:
bottom nav will be hidden if one item is visible

Nikunjsharma0 committed:
bottom nav will be hidden if one item is visible

mikedawson commented:
To be modified: GetDefaultDestinationUseCase to choose where to go based on what is enabled. This can be changed to using a suspend function, so it run a database query and get the site entity from database.

Nikunjsharma0 committed:
BottomNavPermissionConstants.kt removed

Nikunjsharma0 committed:
unused import removed

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
changes in GetDefaultDestinationUseCase

Nikunjsharma0 committed:
login issue fix

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
commit

Nikunjsharma0 committed:
when storage clear app wrong screen open issue fixed

poojaustad committed:
maestro test fix- added -back command

Nikunjsharma0 committed:
error showing if all bottom nav toggle is off

Nikunjsharma0 committed:
updated SiteEditViewModel

Nikunjsharma0 committed:
updated SiteDetailScreen

poojaustad committed:
test updated as per latest design 25/06

poojaustad committed:
test update-007_006

Nikunjsharma0 committed:
primary merged

Nikunjsharma0 committed:
issue fixed

Nikunjsharma0 committed:
added fieldsEnabled in AddAccountSelectNewOrExistingScreen

poojaustad committed:
cypress test flow updated

poojaustad committed:
maestro test flow  and description updated

Nikunjsharma0 committed:
Merge branch 'primary' into primary-bottomnav-visiblitysetting

# Conflicts:
#	core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/signup/SignUpViewModel.kt

Nikunjsharma0 committed:
import issue

mikedawson commented on app-desktop/src/main/java/com/ustadmobile/port/desktop/App.kt:
```diff
@@ -205,6 +211,11 @@ fun main() {
+            val currentDb: UmAppDatabase = di.direct.on(accountManager.activeLearningSpace)
+                .instance(tag = DoorTag.TAG_DB)
+
+            val currentSite by currentDb.siteDao().getSiteAsFlow().collectAsState(initial = null)
```
This doesn't handle when the learning space changes. You need to use the flow from AccountManager to get the current learning space. 

e.g. if you use something like:

var currentSite: Site? by mutableStateOf(null)

LaunchedEffect(Unit) {
   accountManager.currentSession.map {
     it.learningSpace
  }.distinct(). collectLatest {
      val db = di.on(session.learningSpace).instance<Database>
      db.siteDao.collectSiteAsFlow().collect {
           currentSite = it
      }
   }
}

mikedawson commented on app-react/src/jsMain/kotlin/com/ustadmobile/mui/components/Sidebar.kt:
```diff
@@ -1,52 +1,60 @@
 import com.ustadmobile.mui.common.Sizes
 import dev.icerock.moko.resources.StringResource
 import js.objects.jso
 import mui.icons.material.Chat
```
As per web versin notes: when importing icons always as as NameIcon e.g. as was done for school otherwise IDE performance can be very problematic

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/site/edit/SiteEditViewModel.kt:
```diff
@@ -158,6 +175,16 @@ class SiteEditViewModel(
             }
         }
     }
+    fun onTogglePermission(flag: Long) {
```
should be onToggleNavigationItem or something like this - these are not permission flags even though they follow the same pattern

mikedawson replied:
This function should call onEntityChanged; which updates the saved state handle to ensure state is saved if android app destroyed, web user goes back/forth, etc.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/login/LoginViewModel.kt:
```diff
@@ -110,12 +110,21 @@ class LoginViewModel(
+        val getDefaultDestinationUseCase: GetDefaultDestinationUseCase =
+            di.on(LearningSpace(serverUrl)).direct.instance()
+        viewModelScope.launch {
+            if (savedStateHandle[UstadView.ARG_NEXT].equals(PersonListViewModel.DEST_NAME_HOME)){
```
I see no reason why PersonListViewModel needs special treatment here- this is a very weird if statement

mikedawson replied:
This is treating the symptom, not the cause.

mikedawson replied:
This is asynchronous code - and it is creating a race condition (automated test might go faster, and next destination would not be correct in that case). This needs to be refactored to avoid a race condition.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/signup/SignupEnterUsernamePasswordViewModel.kt:
```diff
@@ -134,6 +137,12 @@ class SignupEnterUsernamePasswordViewModel(
 
                 )
             }
+            if (savedStateHandle[UstadView.ARG_NEXT].equals(PersonListViewModel.DEST_NAME_HOME)){
```
This is copy/pasted duplicated logic (also found in loginviewmodel) - which clearly goes against Do not repeat yourself : https://github.com/UstadMobile/UstadMobile/blob/primary/PULLREQUEST-CHECKLIST.md#very-important (from the very important top part of pull request checklist)

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/signup/SignupEnterUsernamePasswordViewModel.kt:
```diff
@@ -206,6 +215,9 @@ class SignupEnterUsernamePasswordViewModel(
             _uiState.value.person?.shallowCopy {
                 this.firstNames = firstName
                 this.lastName = lastName
+                dateOfBirth = savedStateHandle[ARG_DATE_OF_BIRTH]?.toLong()?:0L
```
This belongs in the creation of the entity - it is an initial default being set.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/signup/SignUpViewModel.kt:
```diff
@@ -130,6 +135,14 @@ class SignUpViewModel(
 
     init {
+        viewModelScope.launch {
+            if (savedStateHandle[UstadView.ARG_NEXT].equals(PersonListViewModel.DEST_NAME_HOME)){
```
As per other comments - the cause should be addressed, we should not copy/paste this if statement to address a symptom into three files.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/site/edit/SiteEditViewModel.kt:
```diff
@@ -32,10 +38,13 @@ data class SiteEditUiState(
+    val bottomNavToggleError: String? = null,
     val registrationEnabledError: String? = null,
     val currentSiteTermsLang: UstadMobileSystemCommon.UiLanguage = uiLangs.first(),
+    val permissionLabels: List<Pair<StringResource,Long>> = emptyList()
```
These aren't permission labels even though they use the same pattern - call it flagLabels.

Nikunjsharma0 committed:
pull request comment fixes

Nikunjsharma0 committed:
testing

mikedawson committed:
Add comment on RedirectViewModel .

Nikunjsharma0 committed:
changes in login for GetDefaultDestinationUseCase

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/login/LoginViewModel.kt:
```diff
@@ -110,12 +110,18 @@ class LoginViewModel(
+        val getDefaultDestinationUseCase: GetDefaultDestinationUseCase =
+            di.on(LearningSpace(serverUrl)).direct.instance()
+        viewModelScope.launch {
+            nextDestination = savedStateHandle[UstadView.ARG_NEXT] ?:
```
This is a race condition - no gaurantee that it will be ready first. You could use CompleteableDeferred.

Nikunjsharma0 committed:
worked on pull request comment

mikedawson committed:
Merge branch 'primary' into primary-bottomnav-visiblitysetting

mikedawson committed:
Fix handling of using the site database within React.

The database and site flow was being retrieved in the UstadScreens component without using memo or state; so each render resulted in a new flow instance. Each new flow instance changed the state, resulting in infinite re-renders.

poojaustad committed:
test code updated- removed maxretries

poojaustad committed:
test code updated
